### PR TITLE
[GEOT-6347] Layer creation using REST API very slow in a PostGIS database with thousands of tables

### PR DESCRIPTION
**Pull request by imranrajjad**
Actual Reported Issue
https://osgeo-org.atlassian.net/browse/GEOS-7533?oldIssueView=true

Geotools JIRA
https://osgeo-org.atlassian.net/browse/GEOT-6347

I propose to put commonly used sqltypes inside overrides map and have the behavior inside JDBCDataStore controllable through some param, that if all sqlTypeName have been found through dialect, it should not bother the driver for querying for datatypes inside underlying Database.

### CODE REVIEW

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java [hunk 1]
```diff
@@ -3228,12 +3229,17 @@ private Integer findVarcharColumnLength(AttributeDescriptor att) {
             if (sqlTypeName != null) {
                 sqlTypeNames[i] = sqlTypeName;
             }
-        }
+        }        
+        //Step 2 Geos-7533 make below two lines controllable from hint or param
+        //if all sql type names have been found in dialect dont
+        //go to database GEOS-7533
+        boolean allTypesFound = !ArrayUtils.contains(sqlTypeNames, null);        
+        if(allTypesFound) return sqlTypeNames;
```
Comment (nmco) [inline modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: Space between `ìf`` and ``(`` missing, remember to do a full build to apply automatic formatting.

Comment (nmco) [reply modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: I would add a log messages (level warning) if the type doesn't exist and we need to go to the database.

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/PostGISDialect.java [hunk 2]
```diff
@@ -1114,6 +1115,8 @@ public void registerSqlTypeNameToClassMappings(Map<String, Class<?>> mappings) {
 
     @Override
     public void registerSqlTypeToSqlTypeNameOverrides(Map<Integer, String> overrides) {
+        //Step 1 Geos-7533 put the types in dialect
+        overrides.put(4, "int4");
```
Comment (nmco) [inline modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/PostGISDialect.java]: Please use ``Types.INTEGER`` instead of ``4``.

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/PostGISDialect.java [hunk 3]
```diff
@@ -1363,7 +1366,7 @@ public void postDropTable(String schemaName, SimpleFeatureType featureType, Conn
             throws SQLException {
         Statement st = cx.createStatement();
         String tableName = featureType.getTypeName();
-
+
```
Comment (nmco) [inline modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/PostGISDialect.java]: Unrelated formatting change.

**Comment by aaime** [comment]
I don't see the PR well on my pho e, but dialects are long lived, do o e Wat would be to cache the resulting type map at the first call in a non static field, if people add extra types they will have to reload GS/rebuild the store, does not seem like a big deal

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java [hunk 4]
```diff
@@ -3229,6 +3230,10 @@ private Integer findVarcharColumnLength(AttributeDescriptor att) {
                 sqlTypeNames[i] = sqlTypeName;
             }
         }
+        // Geos-7533 if all sql type names have been found in dialect dont
```
Comment (aaime) [inline modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: Please refer to Geotools tickets, not geoserver ones

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java [hunk 5]
```diff
@@ -3285,6 +3290,9 @@ private Integer findVarcharColumnLength(AttributeDescriptor att) {
 
                     if (sqlType == sqlTypes[i]) {
                         sqlTypeNames[i] = sqlTypeName;
+                        // Geos-7533
```
Comment (aaime) [inline modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: Please refer to Geotools tickets, not geoserver ones

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java [hunk 6]
```diff
@@ -3285,6 +3290,9 @@ private Integer findVarcharColumnLength(AttributeDescriptor att) {
...
+                        // cache the sqlType which required going to database for sql types
+                        getSqlTypeToSqlTypeNameOverrides().put(sqlType, sqlTypeName);
```
Comment (aaime) [inline modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: Nope, this is not all right. The getSqlTypeToSqlTypeNameOverrides is public with a specific semantic defined in its javadoc, what you are registering here is not an override any longer.
I suggest you create a separate map that acts as a cache instead. Also, if you plan to update it this way, make it a concurrent hash map, otherwise code in production might experience a concurrent modification exception.

Comment (imranrajjad) [reply modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: we already have two hashmaps inside the implementation

would it be acceptable that if this method is made a synchronized method, so that if unseen fields are being fetched from database from user1, user2 can wait for the only one time. 

For hash map, if I use the main hashamp instead of the hashmap behind getSqlTypeToSqlTypeNameOverrides() will that do?

**Comment by nmco** [comment]
@imranrajjad what's the status of this?

**Comment by imranrajjad** [comment]
I have addressed @aaime `s requested change

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/PostGISDialect.java (hunk 2 above)
Comment (imranrajjad) [reply modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/PostGISDialect.java]: hard coding removed, this is addressed

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/PostGISDialect.java (hunk 3 above)
Comment (imranrajjad) [reply modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/PostGISDialect.java]: fixed

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java (hunk 5 above)
Comment (imranrajjad) [reply modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: fixed

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java (hunk 6 above)
Comment (imranrajjad) [reply modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: fixed

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java (hunk 4 above)
Comment (imranrajjad) [reply modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: changed to relevant GEOT, fixed

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java (hunk 1 above)
Comment (imranrajjad) [reply modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: I will add the warning message and push changes 👍

Comment (imranrajjad) [reply modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: fixed

**Comment by nmco** [comment]
Thx @imranrajjad.

**Comment by simboss** [comment]
if I am right this is a fix. Can we backport to 2.16 and 2.15?
